trestle/core/catalog_interface.py

```diff
@@ -362,38 +396,84 @@ def _update_all_controls_in_list(self, controls: List[cat.Control]) -> List[cat.
         for control in controls:
             # first update the control itself by getting it from the dict
-            control = self.get_control(control.id)
-            # no warning given if control not found in dict.  it is assumed to have been removed from the catalog.
-            if control is not None:
-                # then update any controls it contains from the dict
-                if control.controls:
-                    control.controls = self._update_all_controls_in_list(control.controls)
-                new_list.append(control)
+            new_control = self.get_control(control.id)
+            if new_control:
+                # this overrides any sub controls in the control itself
+                # any added sub-controls should add to the skipped list
+                new_control.controls = self._update_all_controls_in_list(as_list(control.controls))
+                new_control.controls = none_if_empty(new_control.controls)
+                new_list.append(new_control)
         return new_list
 
     def _update_all_controls_in_group(self, group: cat.Group) -> None:
         """Given a group of controls, create fresh version pulled from the control dict."""
-        if group.controls:
-            group.controls = self._update_all_controls_in_list(group.controls)
-        if group.groups:
-            new_groups: List[cat.Group] = []
-            for sub_group in group.groups:
-                self._update_all_controls_in_group(sub_group)
-                new_groups.append(sub_group)
-            group.groups = new_groups
+        group.controls = self._update_all_controls_in_list(as_list(group.controls))
+        group.controls = none_if_empty(group.controls)
+        new_groups: List[cat.Group] = []
+        for sub_group in as_list(group.groups):
+            self._update_all_controls_in_group(sub_group)
+            new_groups.append(sub_group)
+        group.groups = none_if_empty(new_groups)
+
+    def _insert_control_in_catalog(self, control_handle: ControlHandle) -> None:
+        """Insert the control into the catalog based on its path."""
+        path = control_handle.path
+        node = self._catalog
+        if path[0] != '':
+            for group_id in path:
+                found_group = None
+                for group in as_list(node.groups):
+                    if group.id == group_id:
+                        found_group = group
+                        break
+                if found_group:
+                    node = found_group
+                else:
+                    raise TrestleError(f'No controls found in catalog for group {group.id}')
+            node.title = control_handle.group_title
+            node.class_ = control_handle.group_class
+        node.controls = as_list(node.controls)
+        try:
+            index = [control.id for control in as_list(node.controls)].index(control_handle.control.id)
+            del node.controls[index]
+        except ValueError:
```

Comment:
I assume this is because of the potential non-existince of the value in the list?
If so, I think it is better to initialize the variable for the list and do the check if the element is in the list rather than trigger and catch the exception i.e.
```
control_ids = [control.id for control in as_list(node.controls)]
index = control_ids.index(control_handle.control.id) if control_handle.control.id in control_ids else None
if index:
   del node.controls[index]
```

Reply:
From the web it seems like the valueerror approach is more standard and pythonic - and it involves only one linear pass as opposed to two - not that it really matters.  How about I make a new list function that uses your code so it's in one place, and I use it.

Reply:
Generally I think it is a bad practice to silence the exceptions, the two linear passes here are not too bad considering that we are dealing with relatively small lists in majority of the cases

Reply:
I created a new generic function in list_utils with no exceptions raised.  It takes a list, a value, and a lambda for getting the value of an element.